### DISCUSSION TITLE
Single row access only on a table

### DISCUSSION THREAD

**Discussion by entinio** [supabase/supabase#6333] [Questions] [answered]
I have a table with multiple rows. I don't want this table to be public but a search on a single row to be public.

I created a function : 

```
CREATE OR REPLACE FUNCTION get_company(ssdom TEXT) RETURNS RECORD AS $$
  DECLARE res RECORD;
  BEGIN
    SELECT * INTO res FROM companies_country WHERE id_name = ssdom;
    RETURN res;
  END;
$$ LANGUAGE plpgsql;
```

To make the table private, I created a role and granted SELECT on this table to this role. Unfortunately, I can't use the function through a RPC call with that specific role only.

How could I do that? If I provide a JWT associated with the new role, then this JWT could be used to see the entire table, which I want to keep private. I only want to provide a specific search on this table with a single row returned.

**Comment by GaryAustin1** [comment, marked as answer]
Normally to do this you would lock down the table with RLS false on select. 
Then in your rpc function declare it security definer and do what ever security checking you want.  
For instance where ssdom = id_name.

**Reply by entinio** [reply]
But if I lock down the table, I guess the RPC function won't be able to access it? (Sorry for the obvious thing I must miss here)

**Comment by GaryAustin1** [comment]
You declare the function “security definer” and it assumes privileges of the creator usually a super user, so bypasses RLS.
